src/message.rs

```diff
@@ -1,59 +1,113 @@
-use crate::util::PeerIdentity;
-use bytes::{Bytes, BytesMut};
-use std::convert::TryFrom;
-use std::string::FromUtf8Error;
+use bytes::Bytes;
+use std::collections::vec_deque::{Iter, VecDeque};
+use std::convert::{From, TryFrom};
+use std::fmt;
+
+#[derive(Debug)]
+pub struct ZmqEmptyMessageError;
+
+impl fmt::Display for ZmqEmptyMessageError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Unable to construct an empty ZmqMessage")
+    }
+}
 
 #[derive(Debug, Clone)]
 pub struct ZmqMessage {
-    pub data: Bytes,
+    frames: VecDeque<Bytes>,
 }
 
-impl From<Bytes> for ZmqMessage {
-    fn from(data: Bytes) -> Self {
-        Self { data }
+impl ZmqMessage {
+    pub fn push_back(&mut self, frame: Bytes) {
+        self.frames.push_back(frame);
     }
-}
 
-impl From<BytesMut> for ZmqMessage {
-    fn from(data: BytesMut) -> Self {
-        data.freeze().into()
+    pub fn push_front(&mut self, frame: Bytes) {
+        self.frames.push_front(frame);
+    }
+
+    pub fn iter(&self) -> Iter<'_, Bytes> {
+        self.frames.iter()
+    }
+
+    pub(crate) fn pop_front(&mut self) -> Option<Bytes> {
+        self.frames.pop_front()
+    }
+
+    pub fn len(&self) -> usize {
+        self.frames.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.frames.is_empty()
+    }
+
+    pub fn get(&self, index: usize) -> Option<&Bytes> {
+        self.frames.get(index)
+    }
+
+    pub fn to_vec(&self) -> Vec<Bytes> {
```

Comment:
IMHO this method should take self by value. This will allow to avoid cloning data. Message can be consumed instead

Reply:
I think I made the change you suggested.  It appears clippy doesn't like it though.  It wants to_xxx to take self by reference.

Reply:
I guess they should be named `into_vec` to make clippy happy.

Reply:
Overall LGTM to me through. I think we can merge this tomorrow

Reply:
I will change the function names.  Do you want me to rebase and squash the commits before you merge?

Reply:
Would be great if you cood rebase to make commit history nicer.
Thank you for your contribution and patience.

Reply:
I will work on getting it cleaned up tomorrow.